<a href="https://colab.research.google.com/github/thimaporn-r/data_viz/blob/main/Pandas102.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import os 
path = '/content/drive/My Drive/1pandas'

In [4]:
covid_file_path = os.path.join(path,'pm-20-jan-2021.csv')
print(covid_file_path)

/content/drive/My Drive/1pandas/pm-20-jan-2021.csv


In [5]:
a = path + '/pm-20-jan-2021.csv'
print(a)

/content/drive/My Drive/1pandas/pm-20-jan-2021.csv


In [6]:
data_covid = pd.read_csv(covid_file_path)
data_covid

No.  ...                                         risk
0          1  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
1          2  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
2          3  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
3          4  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
4          5  ...             คนต่างชาติเดินทางมาจากต่างประเทศ
...      ...  ...                                          ...
12648  12649  ...                             State Quarantine
12649  12650  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12650  12651  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
12651  12652  ...                                Cluster ระยอง
12652  12653  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[12653 rows x 10 columns]

.isnull()

In [7]:
data_covid.isnull()

No.  announce_date  ...  district_of_onset   risk
0      False          False  ...               True  False
1      False          False  ...               True  False
2      False          False  ...              False  False
3      False          False  ...               True  False
4      False          False  ...               True  False
...      ...            ...  ...                ...    ...
12648  False          False  ...              False  False
12649  False          False  ...              False  False
12650  False          False  ...              False  False
12651  False          False  ...              False  False
12652  False          False  ...              False  False

[12653 rows x 10 columns]

.any ค่าความจริงแต่ละคอลัมมา and กัน

In [8]:
data_covid.isnull().any()

No.                      False
announce_date            False
notification_date         True
sex                       True
age                       True
nationality               True
province_of_isolation     True
province_of_onset         True
district_of_onset         True
risk                      True
dtype: bool

.all ค่าความจริงแต่ละคอลัมมา and กัน

In [9]:
data_covid.isnull().all()

No.                      False
announce_date            False
notification_date        False
sex                      False
age                      False
nationality              False
province_of_isolation    False
province_of_onset        False
district_of_onset        False
risk                     False
dtype: bool

In [10]:
data_covid.iloc[0,0].isnull()

AttributeError: ignored

In [11]:
data_covid.iloc['No.'][0].isnull()

TypeError: ignored

In [12]:
data_covid.iloc[0,0]

1

In [13]:
data_covid.iloc[:1,0].isnull()

0    False
Name: No., dtype: bool

## HW วนลูป missing .isnull()
สร้างตารางใหม่ที่ค่าใน sex เป็นmissing \\
สรุปว่าทำไม record นั้นๆถึงเป็น missing

# ต่อตารางแกนX แกนY

# แกน Y pd.concat()

In [14]:
datakk = data_covid[data_covid['province_of_onset']=='ขอนแก่น']
datakk.head()

No.  ...                                           risk
180    181  ...                                        สนามมวย
462    463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466  1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970  1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637  2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ

[5 rows x 10 columns]

In [15]:
dataud = data_covid[data_covid['province_of_onset']=='อุดรธานี']
dataud.head()

No.  ...                                         risk
424  425  ...                                      สนามมวย
434  435  ...                                  สถานบันเทิง
471  472  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
883  884  ...                                  สถานบันเทิง
885  886  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้

[5 rows x 10 columns]

In [16]:
datams = data_covid[data_covid['province_of_onset']=='มหาสารคาม']
datams.head()

No.  ...                                         risk
346      347  ...  สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
789      790  ...                                  สถานบันเทิง
6690    6691  ...                            Cluster สมุทรสาคร
10802  10803  ...                            Cluster สมุทรสาคร

[4 rows x 10 columns]

In [17]:
dataMyisan = pd.concat([datakk,dataud,datams])
dataMyisan

No.  ...                                           risk
180      181  ...                                        สนามมวย
462      463  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
1466    1467  ...                  คนไทยเดินทางกลับจากต่างประเทศ
1970    1971  ...  ไปสถานที่ชุมชน เช่น ตลาดนัด สถานที่ท่องเที่ยว
2637    2638  ...                  สัมผัสผู้เดินทางจากต่างประเทศ
2673    2674  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
5948    5949  ...                              Cluster สมุทรสาคร
6082    6083  ...                              Cluster สมุทรสาคร
9333    9334  ...                           อยู่ระหว่างการสอบสวน
9334    9335  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
10610  10611  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11517  11518  ...                           อยู่ระหว่างการสอบสวน
11697  11698  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
11698  11699  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
424      425  ...                                        สนามมวย
434      435  ...                                    สถานบันเทิง
471      472  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
883      884  ...                                    สถานบันเทิง
885      886  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
941      942  ...                                    สถานบันเทิง
1059    1060  ...                  คนไทยเดินทางกลับจากต่างประเทศ
2013    2014  ...                  คนไทยเดินทางกลับจากต่างประเทศ
5810    5811  ...                              Cluster สมุทรสาคร
7433    7434  ...                                  Cluster ระยอง
346      347  ...    สัมผัสใกล้ชิดกับผู้ป่วยยืนยันรายก่อนหน้านี้
789      790  ...                                    สถานบันเทิง
6690    6691  ...                              Cluster สมุทรสาคร
10802  10803  ...                              Cluster สมุทรสาคร

[28 rows x 10 columns]

# แกน X

In [18]:
data_province = data_covid[['No.','announce_date','province_of_onset']]
data_province

No. announce_date province_of_onset
0          1     1/12/2020     กรุงเทพมหานคร
1          2     1/17/2020     กรุงเทพมหานคร
2          3     1/22/2020            นครปฐม
3          4     1/22/2020     กรุงเทพมหานคร
4          5     1/24/2020     กรุงเทพมหานคร
...      ...           ...               ...
12648  12649     1/20/2021            ชลบุรี
12649  12650     1/20/2021             ระยอง
12650  12651     1/20/2021             ระยอง
12651  12652     1/20/2021             ระยอง
12652  12653     1/20/2021               ตาก

[12653 rows x 3 columns]

In [19]:
data_human = data_covid[['No.','age','sex','nationality']]
data_human

No.   age   sex nationality
0          1  61.0  หญิง       China
1          2  74.0  หญิง       China
2          3  73.0  หญิง    Thailand
3          4  68.0   ชาย       China
4          5  66.0  หญิง       China
...      ...   ...   ...         ...
12648  12649  44.0  หญิง    Thailand
12649  12650  52.0  หญิง    Thailand
12650  12651  23.0  หญิง    Thailand
12651  12652  29.0  หญิง    Thailand
12652  12653  22.0  หญิง       Burma

[12653 rows x 4 columns]

ง่ายสุด

In [20]:
full_table1 = data_human.merge(data_province)
full_table1.head()

No.   age   sex nationality announce_date province_of_onset
0    1  61.0  หญิง       China     1/12/2020     กรุงเทพมหานคร
1    2  74.0  หญิง       China     1/17/2020     กรุงเทพมหานคร
2    3  73.0  หญิง    Thailand     1/22/2020            นครปฐม
3    4  68.0   ชาย       China     1/22/2020     กรุงเทพมหานคร
4    5  66.0  หญิง       China     1/24/2020     กรุงเทพมหานคร

In [21]:
data_human2 =data_human.sort_values('No.')
data_human2

No.   age   sex nationality
0          1  61.0  หญิง       China
1          2  74.0  หญิง       China
2          3  73.0  หญิง    Thailand
3          4  68.0   ชาย       China
4          5  66.0  หญิง       China
...      ...   ...   ...         ...
12648  12649  44.0  หญิง    Thailand
12649  12650  52.0  หญิง    Thailand
12650  12651  23.0  หญิง    Thailand
12651  12652  29.0  หญิง    Thailand
12652  12653  22.0  หญิง       Burma

[12653 rows x 4 columns]

In [22]:
data_human2 =data_human.sort_values('age')
data_human2

No.   age   sex nationality
1987    1988  0.10   ชาย    Thailand
11497  11498  0.11   ชาย       Burma
6477    6478  0.11  หญิง    Thailand
1675    1676  0.30   ชาย       Japan
1075    1076  0.40   ชาย    Thailand
...      ...   ...   ...         ...
12635  12636   NaN   NaN    Thailand
12636  12637   NaN   NaN    Thailand
12637  12638   NaN   NaN    Thailand
12638  12639   NaN   NaN    Thailand
12639  12640   NaN   NaN    Thailand

[12653 rows x 4 columns]

In [23]:
full_table2 = data_human2.merge(data_province)
full_table2.head()

No.   age   sex nationality announce_date province_of_onset
0   1988  0.10   ชาย    Thailand      4/4/2020             ระยอง
1  11498  0.11   ชาย       Burma     1/16/2021         สมุทรสาคร
2   6478  0.11  หญิง    Thailand    12/30/2020          เพชรบุรี
3   1676  0.30   ชาย       Japan      4/1/2020     กรุงเทพมหานคร
4   1076  0.40   ชาย    Thailand     3/27/2020      สุราษฎร์ธานี

In [24]:
data_human2_renamed = data_human2.rename(columns={'No.':'patientNumber'})
data_human2_renamed.head()

patientNumber   age   sex nationality
1987            1988  0.10   ชาย    Thailand
11497          11498  0.11   ชาย       Burma
6477            6478  0.11  หญิง    Thailand
1675            1676  0.30   ชาย       Japan
1075            1076  0.40   ชาย    Thailand

In [25]:
full_table3 = data_human2_renamed.merge(data_province,left_on='patientNumber',right_on='No.')
full_table3.head()


patientNumber   age   sex nationality    No. announce_date province_of_onset
0           1988  0.10   ชาย    Thailand   1988      4/4/2020             ระยอง
1          11498  0.11   ชาย       Burma  11498     1/16/2021         สมุทรสาคร
2           6478  0.11  หญิง    Thailand   6478    12/30/2020          เพชรบุรี
3           1676  0.30   ชาย       Japan   1676      4/1/2020     กรุงเทพมหานคร
4           1076  0.40   ชาย    Thailand   1076     3/27/2020      สุราษฎร์ธานี

In [26]:
full_table3 = data_human2_renamed.merge(data_province,how='left''patientNumber',right_on='No.')
full_table3.head()

TypeError: ignored

In [ ]:
data_covid.head()

คุณสมบัติของ padas ก็คือเราสามารถสร้าง คอลัม ใหม่ให้ตาราง df ได้ โดย \\
df['ชื่อ คอลัมใหม่'] = (list ที่มีจำนวนสมาชิกเท่ากับจำนวน record ของ df)


In [ ]:
data_human2_renamed['num'] = range(data_human2_renamed.shape[0])
data_human2_renamed

In [ ]:
data_human2_renamed['datail'] = data_human2_renamed['patientNumber'].map(data_covid.set_index('No.')['risk'])
data_human2_renamed

# HW7

In [27]:
update_sex = data_covid.fillna(value={'sex':'Missing'})
update_sex.head()


No. announce_date  ... district_of_onset                              risk
0    1     1/12/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
1    2     1/17/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
2    3     1/22/2020  ...             เมือง  คนต่างชาติเดินทางมาจากต่างประเทศ
3    4     1/22/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ
4    5     1/24/2020  ...               NaN  คนต่างชาติเดินทางมาจากต่างประเทศ

[5 rows x 10 columns]

In [28]:
update_sex2 = update_sex[update_sex['sex']=='Missing']
update_sex2

No. announce_date  ... district_of_onset               risk
4391    4392    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4392    4393    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4393    4394    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4394    4395    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
4395    4396    12/20/2020  ...             เมือง  Cluster สมุทรสาคร
...      ...           ...  ...               ...                ...
12635  12636     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12636  12637     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12637  12638     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12638  12639     1/20/2021  ...             เมือง  Cluster สมุทรสาคร
12639  12640     1/20/2021  ...             เมือง  Cluster สมุทรสาคร

[2502 rows x 10 columns]

In [49]:
update_sex2.shape

(2502, 10)

Sex มีจำนวนข้อมูลที่เป็น Missing 2502 คน

## nationality

In [30]:
mode_nationality = update_sex2.iloc[:,5]
mode_nationality

4391        Burma
4392        Burma
4393        Burma
4394        Burma
4395        Burma
           ...   
12635    Thailand
12636    Thailand
12637    Thailand
12638    Thailand
12639    Thailand
Name: nationality, Length: 2502, dtype: object

In [31]:
import statistics as st

In [32]:
mode = st.mode(mode_nationality)
mode

'Burma'

In [33]:
 update_nationality = update_sex2['nationality']


In [34]:
 update_nationality[update_sex2['nationality']=='Burma']

4391    Burma
4392    Burma
4393    Burma
4394    Burma
4395    Burma
        ...  
6779    Burma
6780    Burma
6781    Burma
7071    Burma
7074    Burma
Name: nationality, Length: 1366, dtype: object

In [35]:
update_nationality2=update_nationality[update_sex2['nationality']=='Burma']


In [36]:
update_nationality2.shape

(1366,)

## isolation

In [37]:
mode_isolation = update_sex2.iloc[:,6]
mode_isolation

4391     สมุทรสาคร
4392     สมุทรสาคร
4393     สมุทรสาคร
4394     สมุทรสาคร
4395     สมุทรสาคร
           ...    
12635    สมุทรสาคร
12636    สมุทรสาคร
12637    สมุทรสาคร
12638    สมุทรสาคร
12639    สมุทรสาคร
Name: province_of_isolation, Length: 2502, dtype: object

In [38]:
mode = st.mode(mode_isolation)
mode

'สมุทรสาคร'

In [39]:
update_isolation = update_sex2['province_of_isolation']

In [40]:
update_isolation[update_sex2['province_of_isolation']=='สมุทรสาคร']

4391     สมุทรสาคร
4392     สมุทรสาคร
4393     สมุทรสาคร
4394     สมุทรสาคร
4395     สมุทรสาคร
           ...    
12635    สมุทรสาคร
12636    สมุทรสาคร
12637    สมุทรสาคร
12638    สมุทรสาคร
12639    สมุทรสาคร
Name: province_of_isolation, Length: 2313, dtype: object

In [41]:
update_isolation2 = update_isolation[update_sex2['province_of_isolation']=='สมุทรสาคร']
update_isolation2.shape

(2313,)

## onset

In [55]:
mode_onset = update_sex2.iloc[:,7]


In [56]:
mode = st.mode(mode_onset)
mode

'สมุทรสาคร'

In [57]:
update_onset = update_sex2['province_of_onset']

In [58]:
update_onset[update_sex2['province_of_onset']=='สมุทรสาคร']

4391     สมุทรสาคร
4392     สมุทรสาคร
4393     สมุทรสาคร
4394     สมุทรสาคร
4395     สมุทรสาคร
           ...    
12635    สมุทรสาคร
12636    สมุทรสาคร
12637    สมุทรสาคร
12638    สมุทรสาคร
12639    สมุทรสาคร
Name: province_of_onset, Length: 1358, dtype: object

In [59]:
update_onset2=update_onset[update_sex2['province_of_onset']=='สมุทรสาคร']
update_onset2.shape

(1358,)

##district_of_onset

In [50]:
mode_district_of_onset=update_sex2.iloc[:,8]

In [51]:
mode = st.mode(mode_district_of_onset)
print(mode)

เมือง


In [52]:
update_district_of_onset = update_sex2['district_of_onset']

In [53]:
update_district_of_onset[update_sex2['district_of_onset']=='เมือง']

4391     เมือง
4392     เมือง
4393     เมือง
4394     เมือง
4395     เมือง
         ...  
12635    เมือง
12636    เมือง
12637    เมือง
12638    เมือง
12639    เมือง
Name: district_of_onset, Length: 1357, dtype: object

In [54]:
update_district_of_onset2=update_district_of_onset[update_sex2['district_of_onset']=='เมือง']
update_district_of_onset2.shape

(1357,)

##risk

In [42]:
mode_risk = update_sex2.iloc[:,9]


In [43]:
mode = st.mode(mode_risk)
mode

'Cluster สมุทรสาคร'

In [44]:
update_risk = update_sex2['risk']

In [45]:
update_risk[update_sex2['risk']=='Cluster สมุทรสาคร']

4391     Cluster สมุทรสาคร
4392     Cluster สมุทรสาคร
4393     Cluster สมุทรสาคร
4394     Cluster สมุทรสาคร
4395     Cluster สมุทรสาคร
               ...        
12635    Cluster สมุทรสาคร
12636    Cluster สมุทรสาคร
12637    Cluster สมุทรสาคร
12638    Cluster สมุทรสาคร
12639    Cluster สมุทรสาคร
Name: risk, Length: 2258, dtype: object

In [46]:
update_risk2=update_risk[update_sex2['risk']=='Cluster สมุทรสาคร']
update_risk2.shape


(2258,)

ข้อมูลเพศที่เป็นMissingทั้งหมด 2502 คน \\
โดยมีสาเหตุมาจากCluster สมุทรสาคร มากที่สุด 2258 คน \\
รักษาตัวที่สมุทรสาคร จำนวน 2313 คน \\
และพบเป็นสัญชาติเมียนมาร์มากที่สุดถึง 1366 คน \\
จังหวัดที่เริ่มพบว่าตัวเองมีอาการป่วยมากที่สุดคือ สมุทรสาคร จำนวน 1358 คน \\
อำเภอที่เริ่มพบว่าตัวเองมีอาการป่วยมากที่สุด คือ อำเภอเมือง จำนวน 1357 คน \\
Missingที่เกิดขึ้นนี้อาจจะเกิดจากความผิดพลาดของผู้กรอกข้อมูลหรือในขณะนั้นข้อมูลมีจำนวนมากเกินไปจึงทำให้ข้อมูลตกหล่น
และอาจจะเป็นไปได้ว่าผู้ป่วยที่พบว่าเป็นผู้ติดเชื้อในไทยแต่ไปรักษาต่างประเทศจึงทำให้ไม่มีข้อมูลผู้ป่วยคนนั้น




